grails-app/domain/org/pih/warehouse/product/ProductAvailability.groovy

```diff
@@ -60,4 +60,88 @@ class ProductAvailability {
     Boolean isPickable() {
         return (inventoryItem ? inventoryItem.pickable : true) && (binLocation ? binLocation.pickable : true)
     }
+
+    @Override
+    boolean equals(o) {
+        if (this.is(o)) {
+            return true
+        }
+        if (getClass() != o.class) {
+            return false
+        }
+
+        ProductAvailability that = (ProductAvailability) o
+
+        if (binLocation != that.binLocation) {
+            return false
```

Comment:
The order of these should probably matters. I imagine we want id to be first and only, then we compare the objects that are used to define uniqueness (product, inventory item, location, bin location, etc). The quantity, dates, and indexable fields seem like overkill so let me know the justification for including them. Could we ever be in a situation where the unique fields aren't enough? 

We'll need to look at the data but the equals implementation might be simpler if we use those indexable fields (productCode, lotNumber, binLocationName, etc) instead of the objects (product, inventoryItem, binLocation). But I worry about consistency with default values around lotNumber (where we use empty string, NULL, etc) so again we probably need to look at the data there. We don't want two inventory items with different IDs (one with NULL and one with empty string) being unequal when comparing the IDs but equal when comparing the lot number (i.e. assuming there's a case where empty string and NULL are equal).

Reply:
For now I've reverted the `equals`/`hashCode` implementation and I rely on checking `.id`.